client/state/data-layer/third-party/refer/index.js

```diff
@@ -38,12 +55,55 @@ const trackAffiliatePageLoad = action => {
 	);
 };
 
+const onTrackAffiliatePageLoadSuccess = ( action, response ) => {
+	if (
+		'object' !== typeof response ||
+		'object' !== typeof response.body ||
+		'object' !== typeof response.body.data
+	) {
+		debug( 'Unexpected referral response: ', response );
+		return; // Not possible.
+	}
+
+	const responseBody = response.body;
+	const responseData = responseBody.data;
+	const eventProps = pick( responseData, whitelistedEventProps );
+
+	eventProps.status = response.status || '';
+	eventProps.success = responseBody.success || '';
+	eventProps.description = responseBody.message || 'success';
+
+	analytics.tracks.recordEvent( 'calypso_refer_visit_response', eventProps );
+};
+
+const onTrackAffiliatePageLoadError = ( action, error ) => {
+	if (
+		'object' !== typeof error ||
+		'object' !== typeof error.response ||
+		'string' !== typeof error.response.text
+	) {
+		debug( 'Unexpected referral error: ', error );
+		return; // Not possible.
+	}
+
+	const response = error.response;
+	let responseBody = JSON.parse( response.text );
+	responseBody = 'object' === typeof responseBody ? responseBody : {};
```

Comment:
this parsing feels a bit hand-wavey. why is the response coming back as JSON data in the `text` value?

Reply:
Yeah. This API we're calling responds with JSON even on error, but its treated as a failure and the body is not parsed, instead being passed to `onError` where we get hand-wavey.

Instead of `onSuccess` and `onError` it would be nice to put `http()` into a mode whereby all responses are handled by an `onComplete` handler. The `onComplete` handler could look at the response status and response data in greater detail than is typically necessary -- just for cases like this.